File: src/main/java/base/mvp/BasePresenter.java
package base.mvp;

import android.support.annotation.NonNull;
import android.util.Log;

public abstract class BasePresenter<VM extends BaseContract.ViewModel> implements BaseContract.Presenter {

    private static final String TAG = BasePresenter.class.getSimpleName();
    @NonNull
    protected final VM mViewModel;

    protected BasePresenter(@NonNull VM viewModel) {
        mViewModel = viewModel;
    }

    @Override
    public void start() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "started: " + this.getClass().getSimpleName() + "@" + this.hashCode());
        }
    }

    @Override
    public void resume() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "resumed: " + this.getClass().getSimpleName() + "@" + this.hashCode());
        }
    }

    @Override
    public void pause() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "paused: " + this.getClass().getSimpleName() + "@" + this.hashCode());
        }
    }

    @Override
    public void stop() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "stopped: " + this.getClass().getSimpleName() + "@" + this.hashCode());
        }
    }
}

File: src/main/java/base/mvp/BasePresenter.kt
package base.mvp

import android.util.Log

abstract class BasePresenter<VM : BaseContract.ViewModel> (protected val mViewModel: VM) : BaseContract.Presenter {

    override fun start() {
        if (BuildConfig.DEBUG) {
            Log.d(this.javaClass.simpleName, "started: " + javaClass.simpleName + "@" + hashCode())
        }
    }

    override fun resume() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "resumed: " + javaClass.simpleName + "@" + hashCode())
        }
    }

    override fun pause() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "paused: " + javaClass.simpleName + "@" + hashCode())
        }
    }

    override fun stop() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "stopped: " + this.javaClass.simpleName + "@" + hashCode())
        }
    }


        private val TAG = javaClass.simpleName
}
